source/endpoints_eval_util.c

```diff
@@ -0,0 +1,420 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/logging.h>
+#include <aws/common/string.h>
+#include <aws/common/json.h>
+#include <aws/sdkutils/private/endpoints_eval_util.h>
+#include <aws/sdkutils/sdkutils.h>
+
+#include <inttypes.h>
+
+#ifdef _MSC_VER
+#    pragma warning(disable : 4204)
+#    pragma warning(disable : 4706)
+#    pragma warning(disable : 4996)
+#endif
+
+/* 4 octets of 3 chars max + 3 separators + null terminator */
+#define AWS_IPV4_STR_LEN 16
+#define IP_CHAR_FMT "%03" SCNu16
+
+/* arbitrary max length of a region. curent longest region name is 16 chars */
+#define AWS_REGION_LEN 50
+
+struct aws_byte_cursor aws_byte_cursor_from_substring(const struct aws_string *src, size_t start, size_t end) {
+    AWS_PRECONDITION(aws_string_is_valid(src));
+    AWS_PRECONDITION(start < end && end <= src->len);
+
+    return aws_byte_cursor_from_array(aws_string_bytes(src) + start, end - start);
+}
+
```

Comment:
trivial: I don't see this function called from anywhere. Delete it if it's not used

Reply:
will clean up. its victim of refactoring